### PR TITLE
Precision Landing from @ndepal

### PR DESCRIPTION
**Pull request by mhkabir**
@LorenzMeier @ndepal Rebased version of #8160

In addition, I added EKF2 support to select the measurement delay for the landing target estimate, and set it to a default of 5 ms for the IR-Lock. This is necessary for custom setups where the delay is significant.

We also tested this independently with a custom target tracker (which actually takes Gazebo images and runs an algorithm to estimate target pose from it) and everything works pretty good. See demo below.

With respect to documentation, we have fairly complete docs from @ndepal and @hamishwillee here : https://docs.px4.io/en/advanced_features/precland.html. We still need to update the current docs with information about EKF2's auxiliary velocity aiding (added after the docs were created).

### CODE REVIEW

**Comment by mhkabir** [comment]
Wishlist (post-merge) : 
1. Buffer accelerations and range data in `landing_target_estimator` and apply to time-delay-compensated measurement.
2. Enable fully GPS-denied precision landing and position hold. Requires some ECL changes.
3. Ability to enter precision landing from QGC.

**Comment by LorenzMeier** [comment]
Thanks!

**Comment by LorenzMeier** [comment]
@ndepal @mhkabir Build failures still due to clang-tidy. The things it flagged are real code quality issues.

**Comment by LorenzMeier** [comment]
```
/tmp/jenkins/workspace/Firmware_pr-ndepal-precland-CJVXDMTHTMULWYHS75IQQWMWWLV2NBVX6DGLYHOTZ4YWPWQSADXA/src/modules/landing_target_estimator/landing_target_estimator_main.cpp:102:58: error: use nullptr [modernize-use-nullptr,-warnings-as-errors]
                                                 (argv) ? (char *const *)&argv[2] : (char *const *)NULL);
                                                                                                   ^
                                                                                                   nullptr
Suppressed 22 warnings (22 in non-user code).
Use -header-filter=.* to display errors from all non-system headers. Use -system-headers to display errors from system headers as well.
1 warning treated as error
18 warnings generated.
/tmp/jenkins/workspace/Firmware_pr-ndepal-precland-CJVXDMTHTMULWYHS75IQQWMWWLV2NBVX6DGLYHOTZ4YWPWQSADXA/src/modules/landing_target_estimator/KalmanFilter.cpp:45:90: error: the parameter 'covInit' is copied for each invocation but only used as a const reference; consider making it a const reference [performance-unnecessary-value-param,-warnings-as-errors]
KalmanFilter::KalmanFilter(matrix::Vector<float, 2> initial, matrix::Matrix<float, 2, 2> covInit)
                                                                                         ^
/tmp/jenkins/workspace/Firmware_pr-ndepal-precland-CJVXDMTHTMULWYHS75IQQWMWWLV2NBVX6DGLYHOTZ4YWPWQSADXA/src/modules/landing_target_estimator/KalmanFilter.cpp:47:7: error: parameter 'initial' is passed by value and only copied once; consider moving it to avoid unnecessary copies [performance-unnecessary-value-param,-warnings-as-errors]
        init(initial, covInit);
             ^
/tmp/jenkins/workspace/Firmware_pr-ndepal-precland-CJVXDMTHTMULWYHS75IQQWMWWLV2NBVX6DGLYHOTZ4YWPWQSADXA/src/modules/landing_target_estimator/KalmanFilter.cpp:50:87: error: the parameter 'covInit' is copied for each invocation but only used as a const reference; consider making it a const reference [performance-unnecessary-value-param,-warnings-as-errors]
void KalmanFilter::init(matrix::Vector<float, 2> initial, matrix::Matrix<float, 2, 2> covInit)
                                                                                      ^
/tmp/jenkins/workspace/Firmware_pr-ndepal-precland-CJVXDMTHTMULWYHS75IQQWMWWLV2NBVX6DGLYHOTZ4YWPWQSADXA/src/modules/landing_target_estimator/KalmanFilter.cpp:52:7: error: parameter 'initial' is passed by value and only copied once; consider moving it to avoid unnecessary copies [performance-unnecessary-value-param,-warnings-as-errors]
        _x = initial;
             ^
             std::move( )
Suppressed 14 warnings (14 in non-user code).
```

**Comment by mhkabir** [comment]
Thanks for the patch!

File: src/modules/mavlink/mavlink_mission.cpp [hunk 1]
```diff
@@ -1291,6 +1291,7 @@ MavlinkMissionManager::parse_mavlink_mission_item(const mavlink_mission_item_t *
 			mission_item->nav_cmd = NAV_CMD_LAND;
 			// TODO: abort alt param1
 			mission_item->yaw = _wrap_pi(mavlink_mission_item->param4 * M_DEG_TO_RAD_F);
+			mission_item->land_precision = mavlink_mission_item->param2;
```
Comment (dagar) [inline src/modules/mavlink/mavlink_mission.cpp]: @mhkabir @ndepal this isn't part of the mavlink spec.

File: cmake/configs/nuttx_px4fmu-v2_default.cmake [hunk 2]
```diff
@@ -115,6 +115,7 @@ set(config_module_list
 	modules/ekf2
 	#modules/local_position_estimator
 	#modules/position_estimator_inav
+	#modules/landing_target_estimator
```
Comment (dagar) [inline cmake/configs/nuttx_px4fmu-v2_default.cmake]: Try to enable?

File: msg/CMakeLists.txt [hunk 3]
```diff
@@ -121,6 +122,8 @@ set(msg_files
 	vehicle_status_flags.msg
 	vtol_vehicle_status.msg
 	wind_estimate.msg
+	landing_target_pose.msg
+	landing_target_innovations.msg
```
Comment (dagar) [inline msg/CMakeLists.txt]: Sort list.

File: src/modules/logger/logger.cpp [hunk 4]
```diff
@@ -618,6 +618,7 @@ void Logger::add_default_topics()
 	add_topic("vehicle_vision_position");
 	add_topic("vtol_vehicle_status", 200);
 	add_topic("wind_estimate", 200);
+	add_topic("landing_target_pose");
```
Comment (dagar) [inline src/modules/logger/logger.cpp]: sort

File: src/modules/navigator/mission.cpp [hunk 5]
```diff
@@ -769,14 +781,41 @@ Mission::set_mission_items()
 			_mission_item.nav_cmd = NAV_CMD_WAYPOINT;
 			_mission_item.autocontinue = true;
 			_mission_item.time_inside = 0.0f;
+
+		} else if (_mission_item.nav_cmd == NAV_CMD_LAND && _work_item_type == WORK_ITEM_TYPE_DEFAULT) {
+			if (_mission_item.land_precision > 0 && _mission_item.land_precision < 3) {
```
Comment (dagar) [inline src/modules/navigator/mission.cpp]: land_precision > 0 and < 3? Where are these defined?

File: src/modules/ekf2/ekf2_main.cpp [hunk 6]
```diff
@@ -439,7 +443,9 @@ Ekf2::Ekf2():
 	_K_pstatic_coef_y(this, "PCOEF_Y"),
 	_K_pstatic_coef_z(this, "PCOEF_Z"),
 	// non EKF2 parameters
-	_airspeed_disabled(this, "FW_ARSP_MODE", false)
+	_airspeed_disabled(this, "FW_ARSP_MODE", false),
+	_land_tgt_static(this, "LTEST_MODE", false)
```
Comment (dagar) [inline src/modules/ekf2/ekf2_main.cpp]: Can we do this without having EKF2 read the external parameter directly? The module is optional so we can't depend on this parameter being present.

**Comment by dagar** [comment]
Overall looking very good. Needs a rebase and I have a few minor issues/questions.
 - can we fit it into px4fmu-v2_default
 - non standard MAV_CMD_NAV_LAND usage
 - can we publish the mode (moving/stationary) in the landing_target_pose message and avoid having EKF2 read the LTEST_MODE parameter directly? 

There's a lot of incremental history here that's not worth keeping and makes rebasing and viewing blame more difficult than it needs to be (changes coming and going). Consider squashing some of the commits along module boundaries.

File: src/modules/simulator/simulator_mavlink.cpp [hunk 7]
```diff
@@ -429,6 +429,25 @@ void Simulator::handle_message(mavlink_message_t *msg, bool publish)
 
 		break;
 
+	case MAVLINK_MSG_ID_LANDING_TARGET:
+		mavlink_landing_target_t lading_target_mavlink;
```
Comment (mhkabir) [inline src/modules/simulator/simulator_mavlink.cpp]: typo

File: msg/CMakeLists.txt (hunk 3 above)
Comment (mhkabir) [reply msg/CMakeLists.txt]: Done

File: src/modules/logger/logger.cpp (hunk 4 above)
Comment (mhkabir) [reply src/modules/logger/logger.cpp]: Done

File: src/modules/navigator/mission.cpp (hunk 5 above)
Comment (mhkabir) [reply src/modules/navigator/mission.cpp]: Needs mavlink update.

File: src/modules/ekf2/ekf2_main.cpp (hunk 6 above)
Comment (mhkabir) [reply src/modules/ekf2/ekf2_main.cpp]: Done

File: cmake/configs/nuttx_px4fmu-v2_default.cmake (hunk 2 above)
Comment (mhkabir) [reply cmake/configs/nuttx_px4fmu-v2_default.cmake]: Still no :(
`region `flash' overflowed by 849 bytes`

**Comment by mhkabir** [comment]
Addressed comments. Will be squashing and splitting by module soon.

**Comment by mhkabir** [comment]
Rebased and split into commits per-file. Much cleaner history now.

**Comment by LorenzMeier** [comment]
@mhkabir I did reset the author to avoid stealing commits from @ndepal. CI is not passing.

**Comment by mhkabir** [comment]
Thanks! I was trying to find a way to do that :)

**Comment by LorenzMeier** [comment]
@mhkabir could you re-test?

**Comment by LorenzMeier** [comment]
@PX4/testflights Please run normal flight tests - we are just looking for regressions to make sure there are no new issues.

**Comment by LorenzMeier** [comment]
Rebased - @ndepal Could you do one last cross test and we're good to go.

**Comment by ndepal** [comment]
Thanks for all the work, guys!

Tested in SITL, everything still works as expected.

**Comment by hamishwillee** [comment]
@ndepal Given this is now merged, can you confirm that the docs in https://docs.px4.io/en/advanced_features/precland.html are still accurate. 

Can you confirm what version of PX4 this went into - 1.8? (I will remove the "this is not supported" warning)

**Comment by ndepal** [comment]
@hamishwillee Precision land went into v1.7.4beta.

The docs are still accurate. There are also these two PRs: https://github.com/PX4/Firmware/pull/9261, https://github.com/mavlink/qgroundcontrol/pull/6322. I don't know if this means that precision landing can now be selected in QGC or if it's just shown.

**Comment by hamishwillee** [comment]
Thanks @ndepal . I removed the warning in https://github.com/PX4/px4_user_guide/commit/356df6936adf358813ca9f9ec80807712849743f

I didn't add a comment for those PRs. HOpefully it means selected!

**Pull request merged**